roles/metrics-export.lua

```diff
@@ -1,17 +1,274 @@
+                if pathi == pathj then
+                    error("http 'endpoints' must have different paths", 3)
+                end
+            end
+        end
+    end
 end
 
-local function apply()
+local httpd_servers = nil
+
+local function validate_http(conf)
+    if conf ~= nil and type(conf) ~= "table" then
+        error("http configuration must be a table, got " .. type(conf), 2)
+    end
+    conf = conf or {}
 
+    if not is_array(conf) then
+        error("http configuration must be an array, not a map", 2)
+    end
+
+    for _, http_node in ipairs(conf) do
+        validate_http_node(http_node)
+    end
+
+    for i, nodei in ipairs(conf) do
+        local hosti, porti, erri = parse_listen(nodei.listen)
+        assert(erri == nil) -- We should already successfully parse the URI.
+        for j, nodej in ipairs(conf) do
+            if i ~= j then
+                local hostj, portj, errj = parse_listen(nodej.listen)
+                assert(errj == nil) -- The same.
+                if hosti == hostj and porti == portj then
+                    error("http configuration nodes must have different listen targets", 2)
+                end
+            end
+        end
+    end
 end
 
-local function stop()
+local function apply_http(conf)
+    for _, node in ipairs(conf) do
+        if #(node.endpoints or {}) > 0 then
+            local host, port, err = parse_listen(node.listen)
+            if err ~= nil then
+                error("failed to parse URI: " .. err, 2)
+            end
+
+            local httpd = http_server.new(host, port)
+            httpd_servers = httpd_servers or {}
+            table.insert(httpd_servers, httpd)
+
+            for _, endpoint in ipairs(node.endpoints) do
+                httpd:route({
+                    method = "GET",
+                    path = remove_side_slashes(endpoint.path),
+                }, http_handlers[endpoint.format])
+            end
+            httpd:start()
+        end
+    end
+end
 
+local function stop_http()
+    for _, httpd in ipairs(httpd_servers or {}) do
+        httpd:stop()
+    end
+    httpd_servers = nil
 end
 
-return {
-    validate = validate,
-    apply = apply,
-    stop = stop,
+local exports = {
+    ["http"] = {
+        validate = validate_http,
+        apply = apply_http,
+        stop = stop_http,
+    },
 }
+
+M.validate = function(conf)
+    if conf ~= nil and type(conf) ~= "table" then
+        error("configuration must be a table, got " .. type(conf))
+    end
+
+    for export_type, opts in pairs(conf or {}) do
+        if type(export_type) ~= "string" then
+            error("export type must be a string, got " .. type(export_type))
+        end
+        if exports[export_type] == nil then
+            error("unsupported export type '" .. tostring(export_type) .. "'")
+        end
+        exports[export_type].validate(opts)
+    end
+end
+
+M.apply = function(conf)
+    M.stop()
```

Comment:
Well... It definitely will fix some issues like deleting routes, but I don't like the idea that each time the configuration is updated (and, as far as I understand, apply triggers not only on roles_cfg change) we became unavailable. Is it too hard to do things a but more carefully? Missing metrics is a sigh of something unhealthy. In case we introduce something like healthcheck, it won't do us a favor.

Reply:
Thank you. I didn’t think about it from this point of view. Updated.

Reply:
But it's still here

Reply:
Oh, funny, updated!